File: Python/Problems/Problems/problem300.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem300(Problem):
    def test(self) -> bool:
        nums = [10, 9, 2, 5, 3, 7, 101, 18]

        expected = 4

        result = self.lengthOfLIS(nums)

        print(result)

        return result == expected

    def lengthOfLIS(self, nums: List[int]) -> int:
        subs = []
        for num in nums:
            l = 0
            r = len(subs) - 1
            mid = 0
            while l <= r:
                mid = (l + r) // 2
                if num > subs[mid]:
                    l = mid + 1
                else:
                    r = mid - 1
            if l == len(subs):
                subs.append(num)
            else:
                subs[l] = num
        return len(subs)

File: C++/Problems/Problems/problem_300.cpp
#include "problem_300.h"

bool problem_300::test() {
    std::vector nums  { 10, 9, 2, 5, 3, 7, 101, 18 };

    int expected = 4;

    auto result = lengthOfLIS(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_300::lengthOfLIS(std::vector<int> &nums) {
    std::vector<int> subs;
    for (auto num : nums)
    {
        int l = 0;
        int r = subs.size() - 1;
        int mid = 0;
        while (l <= r)
        {
            mid = (l + r) / 2;
            if (num > subs[mid])
                l = mid + 1;
            else
                r = mid - 1;
        }
        if (l == subs.size())
            subs.push_back(num);
        else
            subs[l] = num;
    }

    return subs.size();
}
